cirq/ops/linear_operator.py

```diff
@@ -0,0 +1,316 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple, Union
+import abc
+
+import numpy as np
+import scipy.linalg
+
+from cirq import protocols
+from cirq.linalg import operator_spaces
+from cirq.ops import gate_features
+
+
+class AbstractLinearOperator(metaclass=abc.ABCMeta):
+    """General linear operator on a finite-dimensional vector space.
+
+    AbstractLinearOperators encompass both quantum gates and other (not
+    necessarily unitary) linear operators. Subclasses choose one or more
+    internal representations by overriding some of the abstract methods
+    below. This class in turn equips then with basic linear algebraic
+    operations.
+    """
+    @abc.abstractmethod
+    def _n_dim_(self) -> Optional[int]:
+        """Returns the number of dimensions of the space self acts on."""
+        pass
+
+    @abc.abstractmethod
+    def _matrix_(self) -> Optional[np.ndarray]:
```

Comment:
I'm thinking this should be a protocol instead of an abstract method. I can imagine us wanting to support representations of linear operators that are not committed to a particular basis.